### ISSUE TITLE
[K-Bug] Rule containing Unicode token is not correctly kompiled

### ISSUE THREAD

**Issue by tothtamas28** [runtimeverification/k#3086]
### What component is the issue in?

Front-End

### Which command

- [X] kompile
- [ ] kast
- [ ] krun
- [ ] kprove
- [ ] kprovex
- [ ] ksearch

### What K Version?

v5.5.46-0-gb2a816206d

### Operating System

Linux

### K Definitions (If Possible)

```k
// emoji.k

module EMOJI-SYNTAX
    syntax Emoji ::= r"🙁|🙂" [token]
endmodule

module EMOJI
    imports EMOJI-SYNTAX
    rule <k> 🙁 => 🙂 ... </k>
endmodule
```

#### Haskell backend does not kompile rule correctly

```
$ kompile emoji.k --backend haskell
```

Then in `compiled.txt`:

```
rule `<generatedTop>`(`<k>`(``#token("\U0001f641 =","Emoji")=>#token(" ...","Emoji")``~>_DotVar1),_DotVar0) requires #token("true","Bool") ensures #token("true","Bool") [...]
```

Accordingly in `definition.kore`:

```
alias rule0LHS{}(SortGeneratedCounterCell{},SortK{}) : SortGeneratedTopCell{}
where rule0LHS{}(Var'Unds'DotVar0:SortGeneratedCounterCell{},Var'Unds'DotVar1:SortK{}) :=
  \and{SortGeneratedTopCell{}} (
    \top{SortGeneratedTopCell{}}(), Lbl'-LT-'generatedTop'-GT-'{}(Lbl'-LT-'k'-GT-'{}(kseq{}(inj{SortEmoji{}, SortKItem{}}(\dv{SortEmoji{}}("\U0001f641 =")),Var'Unds'DotVar1:SortK{})),Var'Unds'DotVar0:SortGeneratedCounterCell{})) []

axiom{} \rewrites{SortGeneratedTopCell{}} (
  rule0LHS{}(Var'Unds'DotVar0:SortGeneratedCounterCell{},Var'Unds'DotVar1:SortK{}),
  \and{SortGeneratedTopCell{}} (
    \top{SortGeneratedTopCell{}}(), Lbl'-LT-'generatedTop'-GT-'{}(Lbl'-LT-'k'-GT-'{}(kseq{}(inj{SortEmoji{}, SortKItem{}}(\dv{SortEmoji{}}(" ...")),Var'Unds'DotVar1:SortK{})),Var'Unds'DotVar0:SortGeneratedCounterCell{})))
  [...]
```
#### LLVM backend fails to kompile definition
```
$ kompile emoji.k --verbose
Parse command line options                                   =  0.009s
Importing: Source(Auto imported prelude)
Importing: Source(path/to/k/k-distribution/target/release/k/include/kframework/builtin/prelude.md)
Importing: Source(path/to/k/k-distribution/target/release/k/include/kframework/builtin/kast.md)
Importing: Source(path/to/k/k-distribution/target/release/k/include/kframework/builtin/domains.md)
Importing: Source(path/to/emoji.k)
Outer parsing [61 modules]                                   =  0.599s
Parse configurations [0/3 declarations]                      =  0.259s
  New scanner: EMOJI-RULE-CELLS                              =  0.139s
Parse rules [0/73 rules]                                     =  0.174s
Validate definition                                          =  0.117s
Apply compile pipeline                                       =  0.280s
Kompile to kore                                              =  0.009s
Save to disk                                                 =  0.030s
  Print definition.kore                                      =  0.060s
  Write decision tree                                        =  0.613s
  ┌Executing in path/to/emoji-kompiled: llvm-kompile definition.kore dt main -- -o interpreter
/usr/bin/opt-12: tmp.qf8GSyckHL:8554:1: error: expected top-level entity
\
^
[Error] Critical: llvm-kompile returned nonzero exit code: 1
Examine output to see errors.
```

### Expected Results

K should correctly parse and kompile rules containing Unicode tokens.

**Comment by Baltoli** [comment]
I'm not sure exactly what's causing the LLVM backend to break here; something in the broken unicode parsing from the frontend seems to cause LLVM to incorrectly print a trailing slash. I will revisit the backend treatment once the frontend parsing problem is resolved and we're sure the generated KORE is sensible.

**Comment by radumereuta** [comment]
There is something definitely broken in the front-end parsing regarding unicode.
This parsed term should not look like this: `#token("\U0001f641 =","Emoji")`
Things to check: 
1. the outer syntax parsed correctly: https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/parser/ParserUtils.java#L253
2. the inner scanner has been generated correctly: https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java#L183.
3. the inner parser handles tokens correctly: https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java#L286


Most likely, the problem is with step 2.
You can look in the `*-kompiled` directory and notice that `parsed.txt` already contains the wrong output.

**Comment by radumereuta** [comment]
After a bit of debugging with @Scott-Guest we found an inconsistency between how the [`flex`](https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/parser/inner/kernel/EarleyParser.java#L333) scanner calculates the length of a Unicode character to 4 bytes and how [Java](https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/parser/inner/kernel/EarleyParser.java#L343) thinks of the same character as length 1.

**Comment by radumereuta** [comment]
Another way to get the proper token is to look at `data.words[x].value` and just match the start and end location, but the value matched by the scanner.

**Comment by Scott-Guest** [comment]
To elaborate, the root cause here is that Flex is an "8-bit clean" scanner, treating the emoji characters as 4 separate bit-octets rather than a single character. 

That is, when we write the Flex pattern `🙁|🙂` it is actually treated as `0xf0 0x9f 0x99 0x81 | 0xf0 0x9f 0x99 0x82`. This is fine for tokenization because an input `🙁` will be treated as `0xf0 0x9f 0x99 0x81` in the same way, and thus still match the pattern. 

However, the issue arises in [https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/parser/inner/kernel/EarleyParser.java#L343](https://github.com/runtimeverification/k/blob/master/kernel/src/main/java/org/kframework/parser/inner/kernel/EarleyParser.java#L343) when we take the (correctly) tokenized string and try to produce a `#token` Constant. Specifically, we index into a Java-string (which treats `🙁` as  single character) using the Flex-output location information (which treats `🙁` as 4 characters).

To fix this, we can just convert the `String` to a `byte[]` before indexing.

**Comment by Baltoli** [comment]
@dwightguth recommends checking that the read and write encodings for the KORE parser / LLVM writer are the same. Also look at how string literals are processed

**Comment by Scott-Guest** [comment]
Closing this because the main issues have been addressed by runtimeverification/llvm-backend#737 and #3341, and this specific test case passes now.

However, note that if this test (and more generally, any K-file) used a Unicode character in the range `U+007F` to `U+00FF` it would still fail due to the issue tracked here: https://github.com/runtimeverification/k/issues/3344#issuecomment-1512064546
